server/eclipse-project/src/main/java/us/freeandfair/corla/endpoint/CVRExportImport.java

```diff
@@ -206,55 +217,150 @@ private void parseFile(final Response the_response, final UploadedFile the_file)
-                                                      AuditReason.TIED_CONTEST, 
-                                                      AuditType.NOT_AUDITABLE);
-        dosdb.updateContestToAudit(cta);
-        result = result + 1;
+    while (!success && retries < DOS_DASHBOARD_UPDATE_RETRIES) {
+      try {
+        retries = retries + 1;
+        Main.LOGGER.debug("updating DoS dashboard, attempt " + retries + 
+                          ", county " + the_county.id());
+        Persistence.beginTransaction();
+        final Set<CountyContestResult> contest_results = 
+            CountyContestResultQueries.forCounty(the_county);
+        final DoSDashboard dosdb = Persistence.getByID(DoSDashboard.ID, DoSDashboard.class);
+
+        for (final CountyContestResult ccr : contest_results) {
+          if (ccr.minMargin() == 0) { 
+            // this is a tied contest
+            final ContestToAudit cta = new ContestToAudit(ccr.contest(), 
+                                                          AuditReason.TIED_CONTEST, 
+                                                          AuditType.NOT_AUDITABLE);
+            dosdb.updateContestToAudit(cta);
+            result = result + 1;
+          }
+        }
+        Persistence.commitTransaction();
+        success = true;
+      } catch (final PersistenceException e) {
+        // something went wrong, let's try again
+        if (Persistence.canTransactionRollback()) {
+          try {
+            Persistence.rollbackTransaction();
+          } catch (final PersistenceException ex) {
+            // not much we can do about it
+          }
+        }
+        result = 0;
+        // let's give other transactions time to breathe
+        try {
+          Thread.sleep(calculateTransactionSleep(the_county, retries));
+        } catch (final InterruptedException ex) {
+          // it's OK to be interrupted
+        }
       }
     }
-
+    // we always need a transaction running
+    Persistence.beginTransaction();
+    if (!success) {
+      transactionFailure(the_response, 
+                         "could not update DoS dashboard for county " + the_county.id() + 
+                         " tied contests after " + DOS_DASHBOARD_UPDATE_RETRIES + " retries");
+      // we have to halt manually because a transaction failure doesn't halt
+      halt(the_response);
+    }
+    Main.LOGGER.info("updated DoS dashboard for county " + the_county.id() + 
+                     " tied contests in " + retries + " tries");
     return result;
   }
   
   /**
+   * Calculates a delay, in milliseconds, to sleep before retrying a transaction.
+   * 
+   * @param the_county The county to retry for.
+   * @param the_retries The number of retries so far.
+   */
+  private long calculateTransactionSleep(final County the_county, final int the_retries) {
+    final int county_id_mod = 4;
+    final int retry_fraction = 2;
+    
+    return (the_county.id() % county_id_mod + the_retries / retry_fraction) * 
```

Comment:
I take it that the `%` is just to add jitter to the backoff?

Reply:
Yup